src/tdyutils.c

```diff
@@ -616,3 +641,128 @@ PetscErrorCode ComputeInverseOf3by3Matrix(PetscReal a[9], PetscReal inv_a[9]) {
   PetscFunctionReturn(0);
 }
 
+/* -------------------------------------------------------------------------- */
+/// Performs a LU decomposition of a matrix
+///
+/// The function is from "Numerical Recipes in C"
```

Comment:
Numerical Recipes has horrible licensing conditions, so I don't think we want to paste their stuff into TDycore. I think we should rely on LAPACK here instead. I'm sure we're building it as part of PETSc.

Reply:
Is it still a violation of license if the code is not a copy/paste but written looking at the code provided in the book?

Reply:
Yes: http://numerical.recipes/licenses/redistribute.html

Reply:
There was a reason why I implemented the code from NR instead of using LAPACK, but I can't recall it. It might be that I needed LU of NR to match PFLOTRAN's implementation. I will test out LAPACK.

Reply:
How big are the matrices here? PETSc has some unrolled block kernels and of course you can use LAPACK. I agree that NA is problematic, both on licensing and reliability.
https://www.stat.uchicago.edu/~lekheng/courses/302/wnnr/nr.html

Reply:
These are 4x4 matrices. I was using NA because of the discussion that started the following email thread https://groups.google.com/g/tdycores-dev/c/TXdbtr7Oo_I/m/zgDWAdFmCAAJ. I'm now working on your suggested solution of using transformation on that email thread.

Reply:
@jedbrown's suggested [variable transformation](https://github.com/TDycores-Project/TDycore/pull/221/files#diff-af758c4df5c826801009d3edcf5374849f379423e4115db5a18dfc8445f9a83cR504) significantly improves the condition number of the matrix and testing shows that cubic polynomial coeffs are computed such that `kr<=1`.